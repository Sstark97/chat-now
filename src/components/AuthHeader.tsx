import Image from "next/image"
import type { AuthHeaderProps } from "@customTypes/components"

/**
 * Este componente es el encargado de mostrar el logo, el título y el subtítulo de la página de autenticación
 * @param {string} title
 * @returns component
 * @example <AuthHeader title="Iniciar sesión" />
 */
const AuthHeader = ({ title }: AuthHeaderProps) => {
    return (
        <section className="w-full flex flex-col items-center">
            <Image src="/logo.png" alt="Logo" width={70} height={70} className="2xl:w-1/12" />
            <h1 className="text-3xl font-bold pt-2 2xl:text-4xl 2xl:pt-3">{title}</h1>
        </section>
    )
}

export default AuthHeader
