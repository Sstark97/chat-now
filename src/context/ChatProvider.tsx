import { createContext, MutableRefObject, useEffect, useRef, useState } from "react"
import { useSession } from "next-auth/react"
import { getFrom } from "@lib/utils/fetcher"
import { API } from "@lib/constants/links"
import { isFormValid } from "@lib/utils/user"
import { equals } from "@lib/utils"
import type { ChildrenProps } from "@customTypes/global"
import type { Context } from "@customTypes/context"
import type { Contacts } from "@customTypes/domain"

const ChatContext = createContext<Context>({} as Context)

/**
 * Este componente es el encargado de proveer el contexto de la aplicación
 * @param {ChildrenProps} { children } - children: componente a mostrar
 * @returns component
 * @example <ChatProvider>children</ChatProvider>
 */
const ChatProvider = ({ children }: ChildrenProps) => {
    const { Provider } = ChatContext
    const ref = useRef<HTMLDivElement>() as MutableRefObject<HTMLDivElement>
    const [error, setError] = useState<boolean>(true)
    const [contacts, setContacts] = useState<Contacts[]>([])
    const [selectedChat, setSelectedChat] = useState<Contacts>({} as Contacts)
    const { status } = useSession()

    useEffect(() => {
        const fetchContacts = async () => {
            const data = await getFrom<Contacts[]>(API.GET_CONTACTS)

            if (equals(data, contacts)) {
                setContacts(data)
            }
        }

        if (status === "authenticated") {
            fetchContacts()
        }
    }, [status, contacts])

    /**
     * Este método es el encargado de comprobar si el formulario es válido
     * @returns void
     * @example handleSetErrorsInForm()
     */
    const handleSetErrorsInForm = () => {
        const haveErrors = !isFormValid(ref.current)
        setError(haveErrors)
    }

    /**
     * Este método es el encargado de recargar los contactos
     * @returns Promise<void>
     * @example await reloadContacts()
     */
    const reloadContacts = async () => {
        const data = await getFrom(API.GET_CONTACTS)
        setContacts(data)
    }

    const handleOpenChat = (id: string) => {
        const chat = contacts.find((contact) => contact.id === id) as Contacts
        setSelectedChat(chat)
    }

    return (
        <Provider
            value={{
                ref,
                error,
                contacts,
                selectedChat,
                handleSetErrorsInForm,
                reloadContacts,
                handleOpenChat,
            }}
        >
            {children}
        </Provider>
    )
}

export { ChatContext, ChatProvider }
