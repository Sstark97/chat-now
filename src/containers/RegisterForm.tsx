import useChatContext from "@hooks/useChatContext"
import useForm from "@hooks/useForm"
import Link from "next/link"
import AuthHeader from "@components/AuthHeader"
import Input from "@components/Input"
import PasswordInput from "@components/PasswordInput"
import Button from "@components/Button"
import Error from "@components/Error"
import AuthProviders from "@containers/AuthProviders"
import { errors } from "@lib/constants/validations"
import { AUTH_BUTTONS, INPUT_REGISTER_PLACEHOLDER } from "@lib/constants/authForms"
import { API, REDIRECT } from "@lib/constants/links"

/**
 * Este componente es el encargado de mostrar el formulario de registro
 * @returns component
 * @example <RegisterForm />
 */
const RegisterForm = () => {
    const { ref } = useChatContext()
    const { action: register, error } = useForm(ref, API.REGISTER, REDIRECT.LOGIN)

    /**
     * Esta función es la encargada de ejecutar la función register
     */
    const handleClickInRegister = async () => {
        await register()
    }

    return (
        <div className="w-[90%] lg:w-2/5">
            <AuthHeader title={"Crea tu cuenta"} />
            <form method="post" className="w-full flex flex-col justify-center items-center">
                <div className="w-10/12 md:w-6/12">
                    <AuthProviders />
                </div>
                <div className="w-10/12 md:w-6/12 flex flex-col items-center">
                    {error ? <Error message={error} /> : <></>}
                    <div className="w-full" ref={ref}>
                        <Input
                            className="w-full mt-5"
                            type="text"
                            placeholder={INPUT_REGISTER_PLACEHOLDER.NAME}
                            name="name"
                            errorManager={errors.name}
                        />
                        <Input
                            className="w-full mt-5"
                            type="email"
                            placeholder={INPUT_REGISTER_PLACEHOLDER.EMAIL}
                            name="email"
                            errorManager={errors.email}
                        />
                        <PasswordInput
                            className="w-full mt-5"
                            placeholder={INPUT_REGISTER_PLACEHOLDER.PASSWORD}
                            location="register"
                            validate
                        />
                        <Button value={AUTH_BUTTONS.REGISTER} action={handleClickInRegister} />
                    </div>
                </div>
            </form>
            <p className="mt-5 text-sm text-center text-secondary_text lg:text-lg">
                ¿Ya tienes una cuenta?{" "}
                <Link href="/login" className="font-bold">
                    Inicia sesión
                </Link>
            </p>
        </div>
    )
}

export default RegisterForm
