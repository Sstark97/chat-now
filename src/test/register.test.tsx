import { render, screen } from "@testing-library/react"
import { ChatProvider } from "@context/ChatProvider"
import AuthHeader from "@components/AuthHeader"
import RegisterForm from "@containers/RegisterForm"

jest.mock("next/router", () => require("next-router-mock"))

describe("Register", () => {
    it("renders a heading", () => {
        const title = "irrelevant title"
        const { container } = render(<AuthHeader title={title} />)

        const heading = screen.getByText(title)

        expect(container).toHaveTextContent(title)
        expect(heading).toBeInTheDocument()
        expect(container).toMatchSnapshot()
    })

    it("render register button in Form", () => {
        render(
            <ChatProvider>
                <RegisterForm />
            </ChatProvider>
        )
        const registerBtn = screen.getByText("Crear cuenta")

        expect(registerBtn).toBeInTheDocument()
    })

    it("check if the button is disables when the form is empty", () => {
        render(
            <ChatProvider>
                <RegisterForm />
            </ChatProvider>
        )

        const registerBtn = screen.getByText("Crear cuenta")

        expect(registerBtn).toHaveAttribute("disabled")
    })
})
