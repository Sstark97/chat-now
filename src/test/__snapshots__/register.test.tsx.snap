// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register renders a heading 1`] = `
<div>
  <section
    class="w-full flex flex-col items-center"
  >
    <img
      alt="Logo"
      class="2xl:w-1/12"
      data-nimg="1"
      decoding="async"
      height="70"
      loading="lazy"
      src="/_next/image?url=%2Flogo.png&w=256&q=75"
      srcset="/_next/image?url=%2Flogo.png&w=96&q=75 1x, /_next/image?url=%2Flogo.png&w=256&q=75 2x"
      style="color: transparent;"
      width="70"
    />
    <h1
      class="text-3xl font-bold pt-2 2xl:text-4xl 2xl:pt-3"
    >
      irrelevant title
    </h1>
  </section>
</div>
`;
